### PR TITLE
Add dmcache input plugin

### PR DESCRIPTION
**Pull request by vlasad**
### Required for all PRs:
- [x] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [x] README.md updated (if adding a new plugin)

### CODE REVIEW

**Comment by sebito91** [comment]
Checking in on this PR, the CircleCI issues weren't fails from the dmcache code but from other components.

**Comment by sparrc** [comment]
yes, it does appear so, I'll slate this for 1.3 and try to get it reviewed soon

File: plugins/inputs/dmcache/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# DMCache Input Plugin
+
+This plugin provide a native collection for dmsetup based statistics for dm-cache
+
+## Configuration
+
+```
+[[inputs.dmcache]]
+## Whether to report per-device stats or not
```
Comment (sparrc) [inline plugins/inputs/dmcache/README.md]: these lines should both be indented by two spaces

File: plugins/inputs/dmcache/dmcache.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type DMCache struct {
+	PerDevice bool `toml:"per_device"`
+	rawStatus func() ([]string, error)
+}
+
+var sampleConfig = `
+	## Whether to report per-device stats or not
```
Comment (sparrc) [inline plugins/inputs/dmcache/dmcache.go]: indent by 2 spaces instead of tabs

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+	if !c.PerDevice {
+		acc.AddFields(metricName, total, nil)
+	}
+
+	return nil
+}
+
+func parseDMSetupStatus(line string) (status map[string]interface{}, err error) {
+	defer func() {
```
Comment (sparrc) [inline plugins/inputs/dmcache/dmcache_linux.go]: why would this function panic? you should write your code to avoid panics rather than write a recover()

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+		if r := recover(); r != nil {
+			status = nil
+			err = r.(error)
+		}
+	}()
+
+	values := strings.Split(line, " ")
```
Comment (sparrc) [inline plugins/inputs/dmcache/dmcache_linux.go]: you might want to use `strings.Fields` here in case of multiple spaces or tabs

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	status = make(map[string]interface{})
+
+	status["device"] = values[0][:len(values[0])-1]
```
Comment (sparrc) [inline plugins/inputs/dmcache/dmcache_linux.go]: all of these should check length of the array to avoid panicking.

Comment (sparrc) [inline plugins/inputs/dmcache/dmcache_linux.go]: or you can check that the array is of a minimum length, and just ignore all metrics if it isn't.

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	status["dirty"] = toInt(values[14])
+	status["blocksize"] = 512
+
+	return status, nil
+}
+
+func toInt(s string) int {
+	i, err := strconv.Atoi(s)
+	if err != nil {
+		panic(err)
```
Comment (sparrc) [inline plugins/inputs/dmcache/dmcache_linux.go]: don't panic, return error

File: plugins/inputs/dmcache/dmcache_test.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
+// +build linux
```
Comment (sparrc) [inline plugins/inputs/dmcache/dmcache_test.go]: I believe you can run these tests on non-linux system as well

File: plugins/inputs/dmcache/dmcache.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+func (c *DMCache) Description() string {
+	return "Provide a native collection for dmsetup based statistics for dm-cache"
+}
+
+func init() {
+	inputs.Add("dmcache", func() telegraf.Input {
+		return &DMCache{
+			PerDevice: true,
+			rawStatus: func() ([]string, error) {
+				out, err := exec.Command("/bin/sh", "-c", "sudo /sbin/dmsetup status --target cache").Output()
```
Comment (sparrc) [inline plugins/inputs/dmcache/dmcache.go]: you need to document that this plugin requires sudo, and that the telegraf user needs to be setup to be able to execute sudo without a password.

You should also document that `sudo /sbin/dmsetup status --target cache` is the full command that telegraf will run for debugging purposes.

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+		data, err := parseDMSetupStatus(s)
+		if err != nil {
+			return err
+		}
+
+		for _, f := range fieldNames {
+			fields[f] = calculateSize(data, f)
+		}
+
+		if c.PerDevice {
```
Comment (sparrc) [inline plugins/inputs/dmcache/dmcache_linux.go]: this if-statement doesn't seem correct to me.

If the user sets per_device = true, then it should report the per-device stats AND total stats.

Is there a reason users would not want total stats? If so it should be a separate config option.

**Comment by vlasad** [comment]
All requested changes were done.

File: plugins/inputs/dmcache/dmcache.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+				if err != nil {
+					return nil, err
+				}
+				if string(out) == "No devices found\n" {
+					return nil, nil
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache.go]: Return an empty slice instead of nil

File: plugins/inputs/dmcache/dmcache.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+				}
+
+				status := strings.Split(string(out), "\n")
+				status = status[:len(status)-1] // removing last empty line
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache.go]: This will panic if there is status is not at least length 1, I suggest you strings.TrimRight(out, "\n") before splitting lines.

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+	return nil
+}
+
+func parseDMSetupStatus(line string) (map[string]interface{}, error) {
+	var err error
+	status := make(map[string]interface{})
+	values := strings.Fields(line)
+	if len(values) < 15 {
+		return nil, errors.New("dmsetup status data have invalid format")
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: Grammar, I suggest "Output from dmsetup could not be parsed"

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	status["device"] = values[0][:len(values[0])-1]
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: If this is trimming the colon on the right, use TrimRight

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+	}
+	status["metadata_total"], err = strconv.Atoi(strings.Split(values[5], "/")[1])
+	if err != nil {
+		return nil, err
+	}
+	status["cache_blocksize"], err = strconv.Atoi(values[6])
+	if err != nil {
+		return nil, err
+	}
+	status["cache_used"], err = strconv.Atoi(strings.Split(values[7], "/")[0])
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: Split once, make sure it return len of 2

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	acc.AddFields(metricName, total, map[string]string{"device": "all"})
+
+	return nil
+}
+
+func parseDMSetupStatus(line string) (map[string]interface{}, error) {
+	var err error
+	status := make(map[string]interface{})
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: You should define a struct for these values, that way less casting is needed later

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+	status["dirty"], err = strconv.Atoi(values[14])
+	if err != nil {
+		return nil, err
+	}
+	status["blocksize"] = 512
+
+	return status, nil
+}
+
+func calculateSize(data map[string]interface{}, key string) (value int) {
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: If I understand this correctly, we are synthesizing a few aggregate values in this function?  Why not just provide the values dmsetup returns, since the user can always calculate these at query time.

File: plugins/inputs/dmcache/dmcache_test.go [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+		"read_hits":     194314240,
+		"read_misses":   92461989888,
+		"write_hits":    21299200,
+		"write_misses":  34397618176,
+		"demotions":     0,
+		"promotions":    1835008,
+		"dirty":         0,
+	}
+	acc.AssertContainsTaggedFields(t, "dmcache", fields, tags)
+}
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_test.go]: The tests need to be made more comprehensive.  Please try to think of some more corner cases and error conditions to test for.  Some things that come to mind is the "No devices found" error and errors from rawOutput().  Try checking the coverage for more ideas.

File: plugins/inputs/dmcache/dmcache.go [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+	"os/exec"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type DMCache struct {
+	PerDevice bool `toml:"per_device"`
+	rawStatus func() ([]string, error)
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache.go]: I feel like the name `rawStatus` is not fitting, since this functions both executes the command and does some processing on the output.

File: plugins/inputs/dmcache/dmcache.go [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+func (c *DMCache) Description() string {
+	return "Provide a native collection for dmsetup based statistics for dm-cache"
+}
+
+func init() {
+	inputs.Add("dmcache", func() telegraf.Input {
+		return &DMCache{
+			PerDevice: true,
+			rawStatus: func() ([]string, error) {
+				out, err := exec.Command("/bin/sh", "-c", "sudo /sbin/dmsetup status --target cache").Output()
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache.go]: Maybe it is required for sudo?

Comment (vlasad) [inline plugins/inputs/dmcache/dmcache.go]: Yes. It's required for sudo.

File: plugins/inputs/dmcache/dmcache_linux.go (hunk 15 above)
Comment (vlasad) [inline plugins/inputs/dmcache/dmcache_linux.go]: But, in this case, function "calculateSize" will requre to use reflection to get value by key.

File: plugins/inputs/dmcache/dmcache_linux.go (hunk 16 above)
Comment (vlasad) [inline plugins/inputs/dmcache/dmcache_linux.go]: #1581 - this is the source of this solution. 
@daviesalex could you comment?

Comment (alexdaviesjump) [inline plugins/inputs/dmcache/dmcache_linux.go]: Thats exactly what is going on. I had a look at it and I agree, these should be do-able at query time without drama (in very old versions of InfluxDB it was a challenge to do everything at query time if you wanted to select if I recall, which is why I think this started off this way).

@vlasad I suggest we remove this. If we did want to do this, the neat way to do it is probably a output plugin in telegraf. 

@sebito91 FYI

Comment (vlasad) [inline plugins/inputs/dmcache/dmcache_linux.go]: removed

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+func parseDMSetupStatus(line string) (map[string]interface{}, error) {
+	var err error
+	parseError := errors.New("Output from dmsetup could not be parsed")
+	status := make(map[string]interface{})
+	values := strings.Fields(line)
+	if len(values) < 15 {
+		return nil, parseError
+	}
+
+	status["device"] = strings.TrimRight(values[0], ":")
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: Still want these in a struct, this separates the responsibilities.  Call it something appropriate that represents the data, maybe CacheStatus.  I get why you did this but I think it will clear up some weirdness in the code, such as taking `device` from fields to add to tags.  You won't need to do type conversions and summing will be more explicit.  This open PR does a good job. #2031

File: plugins/inputs/dmcache/dmcache_test.go [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+	var acc testutil.Accumulator
+
+	err := dmc1.Gather(&acc)
+	require.NoError(t, err)
+
+	tags1 := map[string]string{
+		"device": "cs-1",
+	}
+	fields1 := map[string]interface{}{
+		"device":             "cs-1",
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_test.go]: Device shouldn't be both a tag and a field.

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+	}
+	status["promotions"], err = strconv.Atoi(values[13])
+	if err != nil {
+		return nil, err
+	}
+	status["dirty"], err = strconv.Atoi(values[14])
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: What about the values after dirty, why do we not collect them?

Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: What about the values after dirty, do we need to collect them?

File: plugins/inputs/dmcache/README.md [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+
+`sudo /sbin/dmsetup status --target cache` is the full command that telegraf will run for debugging purposes.
+
+## Configuration
+
+```
+[[inputs.dmcache]]
+  ## Whether to report per-device stats or not
+  per_device = true
+```
```
Comment (sebito91) [inline plugins/inputs/dmcache/README.md]: All of the fields are basically taken from the output of the `dmsetup status --target=cache` output for the given disks. In this case, we want to split all the data into multiple lines per disk we set up.

File: plugins/inputs/dmcache/dmcache.go [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+
+func (c *DMCache) SampleConfig() string {
+	return sampleConfig
+}
+
+func (c *DMCache) Description() string {
+	return "Provide a native collection for dmsetup based statistics for dm-cache"
+}
+
+func dmSetupStatus() ([]string, error) {
```
Comment (sebito91) [inline plugins/inputs/dmcache/dmcache.go]: This is fine, how does one accomplish this on Windows (not a Windows user/fan).

File: plugins/inputs/dmcache/dmcache_test.go (hunk 21 above)
Comment (sebito91) [inline plugins/inputs/dmcache/dmcache_test.go]: that is valid, in the actual output it's not a field but just a tag.

File: plugins/inputs/dmcache/dmcache_linux.go (hunk 22 above)
Comment (sebito91) [inline plugins/inputs/dmcache/dmcache_linux.go]: Not sure what you're asking here, are you questioning why we're collecting the "dirty" blocks?

**Comment by sebito91** [comment]
Here is some output to (hopefully) provide some clarity...

```
[root@hostname ~]# telegraf --test --config /etc/telegraf/telegraf.conf --input-filter dmcache
* Plugin: inputs.dmcache, Collection 1
> dmcache,bu=linux,cls=server,dc=colo,device=vg02-splunk_data_lv,env=production,host=hostname,sr=splunk_indexer,trd=false cache_free=0i,cache_used=1099511627776i,demotions=3265223720960i,dirty=0i,
metadata_free=1060880384i,metadata_used=12861440i,promotions=3265223720960i,read_hits=995134034411520i,read_misses=916807089127424i,write_hits=195107267543040i,write_misses=563725346013184i 1491362011000000000
[root@hostname ~]# dmsetup status --target=cache
vg02-splunk_data_lv: 0 70301769728 cache 8 3140/262144 2048 1048576/1048576 949033802 874335375 186072328 537610402 3113960 3113960 0 1 writethrough 2 migration_threshold 2048000 smq 0 rw -  
```

**Comment by sebito91** [comment]
@danielnelson, @sparrc we've made the proposed changes if you can please give it another look?

File: plugins/inputs/dmcache/dmcache_linux.go (hunk 22 above)
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: I noticed in the sample output included in the tests that there are more values after the column containing "dirty".  Are these metrics important, should we be collecting them, and why?

File: plugins/inputs/dmcache/README.md (hunk 23 above)
Comment (danielnelson) [inline plugins/inputs/dmcache/README.md]: So the output of that command, based on the test cases, is more or less a table of values.  Where did you find out what each column represents?

File: plugins/inputs/dmcache/README.md [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+
+- All measurements have the following tags:
+    - device
+
+### Example Output:
+
+```
+$ ./telegraf --test --config /etc/telegraf/telegraf.conf --input-filter dmcache
+* Plugin: inputs.dmcache, Collection 1
+> dmcache,bu=linux,cls=server,dc=colo,device=vg02-splunk_data_lv,env=production,host=hostname,sr=splunk_indexer,trd=false cache_free=0i,cache_used=1099511627776i,demotions=3265223720960i,dirty=0i,metadata_free=1060880384i,metadata_used=12861440i,promotions=3265223720960i,read_hits=995134034411520i,read_misses=916807089127424i,write_hits=195107267543040i,write_misses=563725346013184i 1491362011000000000
```
Comment (danielnelson) [inline plugins/inputs/dmcache/README.md]: Can you clean out the custom tags you have?

File: plugins/inputs/dmcache/dmcache_linux.go [hunk 26]
```diff
@@ -0,0 +1,180 @@
...
+		fields["write_misses"] = status.writeMisses
+		fields["demotions"] = status.demotions
+		fields["promotions"] = status.promotions
+		fields["dirty"] = status.dirty
+
+		if c.PerDevice {
+			tags := map[string]string{"device": status.device}
+			acc.AddFields(metricName, fields, tags)
+		}
+		aggregateStats(total, fields)
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_linux.go]: Switch to fields at the last moment.  Overall flow should be get the cacheStatus, aggregate, then convert to fields and emit as needed.

File: plugins/inputs/dmcache/dmcache_test.go [hunk 27]
```diff
@@ -0,0 +1,176 @@
...
+		"cs-2: 0 4294967296 cache 8 72352/1310720 128 26/24327168 2409 286 265 524682 0 0 0 1 writethrough 2 migration_threshold 2048 mq 10 random_threshold 4 sequential_threshold 512 discard_promote_adjustment 1 read_promote_adjustment 4 write_promote_adjustment 8",
+	}, nil
+}
+
+var dmc1 = &DMCache{
+	PerDevice:        true,
+	getCurrentStatus: output2Devices,
+}
+
+func TestDMCacheStats_1(t *testing.T) {
```
Comment (danielnelson) [inline plugins/inputs/dmcache/dmcache_test.go]: Can you move the DMCache initialization into the function and rename the tests like `TestWhateverThisTests`.

**Comment by danielnelson** [comment]
I just noticed this, can you get it?
`plugins/inputs/dmcache/dmcache.go:30: undefined: dmSetupStatus`

**Comment by vlasad** [comment]
I cannot get it. 
dmSetupStatus is defined in plugins/inputs/dmcache/dmcache_linux.go:177

**Comment by danielnelson** [comment]
On the non linux builds.

**Comment by dswarbrick** [comment]
I was also working on a dmcache plugin for Telegraf, but put it aside whilst waiting for another plugin I wrote to be accepted, during what seems to have been quite a long moratorium on accepting new plugins.

I took a slightly different approach, and implemented the ioctls so that it does not require os.Exec(). However, it does mean that the telegraf process must be able to open /dev/mapper/control with O_RDWR. If people are interested, I could merge my approach into this plugin.

**Comment by danielnelson** [comment]
@dswarbrick I'll reply over on the issue.
